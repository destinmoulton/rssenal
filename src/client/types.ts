import { List, Map, OrderedMap, Set } from "immutable";
import { Notification } from "react-notification-system";
/** Redux Interfaces */
export interface IDispatch {
    <R>(
        asyncAction: (dispatch: IDispatch, getState: () => IRootStoreState) => R
    ): R;
    <R>(asyncAction: (dispatch: IDispatch) => R): R;
    // (neverAction: (dispatch: Dispatch, getState: () => GetState) => never): never;
    (action: object): void;
    // (action: Thunk): ; // thunks in this app must return a promise
}

export interface IGetState {
    (): IRootStoreState;
}

export interface IAuthAction {
    type: string;
    username?: string;
    password?: string;
    message?: string;
}

export type TEntryID = string;

export interface IEntry {
    _id: TEntryID;
    feed_id: string;
    guid: string;
    title: string;
    link: string;
    creator: string;
    content: string;
    content_snippet: string;
    publish_date: string;
    has_read: boolean;
    feedTitle?: string;
    timeAgo?: string;
}

export interface IEntriesAction {
    type: string;
    entries?: TEntries;
}

export type TEntries = OrderedMap<TEntryID, IEntry>;

/** Feeds Interfaces & Types */

export type TFeedID = string;

export interface IFeed {
    _id: TFeedID;
    title: string;
    folder_id: string;
    description: string;
    link: string;
    creation_date: string;
    last_updated: string;
}

export type TFeeds = OrderedMap<TFeedID, IFeed>;

export interface IFeedsAction {
    type: string;
    feeds?: TFeeds;
    unreadMap?: IFeedsUnreadMap;
}

export type TUnreadMapFeeds = Map<TFeedID, number>;
export type TUnreadMapFolders = Map<TFolderID, number>;
export interface IFeedsUnreadMap {
    entriesCounted: Set<TFeedID>;
    feeds: TUnreadMapFeeds;
    folders: TUnreadMapFolders;
}

/** Filter Interfaces */

export interface IFilter {
    id?: string;
    limit: string;
}
export interface IFilterAction {
    type: string;
    newFilter?: object;
    filterTitle?: string;
    filteredEntries?: TEntries;
}

/** Folders Interfaces & Types */

export type TFolderID = string;

export interface IFolder {
    _id: TFolderID;
    name: string;
    order: number;
}

export type TFolders = OrderedMap<TFolderID, IFolder>;

export interface IFolderAction {
    type: string;
    folders?: TFolders;
}

/** Messages Interfaces */
export interface IMessage {
    message: string;
    level: "error" | "warning" | "info" | "success";
    uid?: number;
    onRemove?: (message: IMessage) => void;
    position?: "br" | "tr" | "tl" | "tc" | "bl" | "bc";
}

export type TMessages = List<IMessage>;

export interface IMessageAction {
    type: string;
    messages: List<IMessage>;
    lastUID?: number;
}

/** Redux Reducer Interfaces */

export interface IReducerStateAuth {
    authenticationError: string;
    isAuthorized: boolean;
    isValidatingToken: boolean;
}

export interface IReducerStateEntries {
    entries: TEntries;
}

export interface IReducerStateFeeds {
    feeds: TFeeds;
    unreadMap: IFeedsUnreadMap;
    isAddingFeed: boolean;
    isUpdatingFeed: boolean;
}

export interface IReducerStateSettings {
    settings: TSettings;
}

export interface IReducerStateFilter {
    filter: IFilter;
    filterTitle: string;
    filteredEntries: TEntries;
}

export interface IReducerStateFolders {
    folders: TFolders;
    hasFolders: boolean;
    isDeletingFolder: boolean;
    isFetchingFolders: boolean;
    isSavingFolder: boolean;
}

export interface IReducerStateMessages {
    messages: TMessages;
    lastUID: 0;
}

export interface IRootStoreState {
    authStore: IReducerStateAuth;
    entriesStore: IReducerStateEntries;
    feedsStore: IReducerStateFeeds;
    filterStore: IReducerStateFilter;
    foldersStore: IReducerStateFolders;
    messagesStore: IReducerStateMessages;
    settingsStore: IReducerStateSettings;
}

/** Settings Interfaces */

export interface ISetting {
    name: string;
    type: string;
    refresh_entries_on_change: boolean;
    value: any;
}

export type TSettings = Map<string, ISetting>;

export interface ISettingsAction {
    type: string;
    settings?: TSettings;
}
