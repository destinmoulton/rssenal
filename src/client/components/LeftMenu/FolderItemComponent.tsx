import { Map, OrderedMap } from "immutable";
import * as React from "react";

import { Icon, SemanticICONS } from "semantic-ui-react";

import ListFeeds from "./ListFeeds";

import {
    IFeed,
    IFolder,
    IFilter,
    TFolderID,
    TFeedID,
    TFeeds
} from "../../interfaces";

export interface IFolderItemMapState {
    filter: IFilter;
    unreadMapGroups: Map<TFolderID, number>;
    feeds: TFeeds;
}

export interface IFolderItemMapDispatch {
    beginDeleteFolder: (folderId: TFolderID) => void;
    changeFilter: (filter: IFilter) => void;
}

interface IFolderItemProps {
    editFeed: (feed: IFeed) => void;
    editFolder: (currentFolder: IFolder) => void;
    folder: IFolder;
}

type TAllProps = IFolderItemProps &
    IFolderItemMapDispatch &
    IFolderItemMapState;

interface IFolderItemState {
    feedsAreVisible: boolean;
    optionsAreVisible: boolean;
}

class FolderItemComponent extends React.Component<TAllProps, IFolderItemState> {
    state: IFolderItemState = {
        feedsAreVisible: false,
        optionsAreVisible: false
    };

    _handleClickDelete = () => {
        const { beginDeleteFolder, folder } = this.props;
        const conf = confirm(
            `Are you sure you want to delete the '${
                folder.name
            }' folder?\nThe feeds will be moved to 'Uncategorized'.`
        );
        if (conf) {
            beginDeleteFolder(folder._id);
        }
    };

    _handleClickEdit = () => {
        this.props.editFolder(this.props.folder);
    };

    _handleClickFolder = () => {
        const { changeFilter, folder } = this.props;

        changeFilter({
            limit: "folder",
            id: folder._id
        });
    };

    _handleHideOptions = () => {
        this.setState({
            optionsAreVisible: false
        });
    };

    _handleShowOptions = () => {
        const { folder } = this.props;

        if (folder._id != "0") {
            this.setState({
                optionsAreVisible: true
            });
        }
    };

    _handleToggleFeedsVisible = () => {
        this.setState({
            feedsAreVisible: !this.state.feedsAreVisible
        });
    };

    _domOptionButtons() {
        return (
            <div className="rss-folders-folderitem-options">
                <Icon
                    name="pencil"
                    color="green"
                    onClick={this._handleClickEdit}
                />
                <Icon
                    name="trash"
                    color="red"
                    onClick={this._handleClickDelete}
                />
            </div>
        );
    }

    _domFolderTitle() {
        const { folder } = this.props;

        return (
            <div
                className="rss-folders-folderitem-title"
                onClick={this._handleClickFolder}
            >
                {folder.name}
            </div>
        );
    }

    _domUnreadCount() {
        const { folder, unreadMapGroups } = this.props;
        let unread = "";
        if (
            unreadMapGroups.has(folder._id) &&
            unreadMapGroups.get(folder._id) > 0
        ) {
            unread = " [" + unreadMapGroups.get(folder._id) + "]";
        }

        return (
            <div
                className="rss-folders-folderitem-unread"
                onClick={this._handleClickFolder}
            >
                {unread}
            </div>
        );
    }

    _domToggleIcon() {
        const { feedsAreVisible } = this.state;

        var toggleFeedsIconClass: SemanticICONS = feedsAreVisible
            ? "caret down"
            : "caret right";

        return (
            <div className="rss-folders-folderitem-icon">
                <Icon
                    name={toggleFeedsIconClass}
                    onClick={this._handleToggleFeedsVisible}
                />
            </div>
        );
    }

    render() {
        const { editFeed, feeds, filter, folder } = this.props;

        const { feedsAreVisible, optionsAreVisible } = this.state;

        let toggleFeedsIcon = null;
        let title = null;
        let unread = null;
        let options = null;
        let listFeeds = null;

        title = this._domFolderTitle();
        if (folder._id !== "all") {
            toggleFeedsIcon = this._domToggleIcon();

            unread = this._domUnreadCount();
            if (feedsAreVisible) {
                listFeeds = (
                    <ListFeeds
                        folderId={folder._id}
                        editFeed={editFeed}
                        feeds={feeds}
                    />
                );
            }

            if (optionsAreVisible) {
                options = this._domOptionButtons();
            }
        }

        let className = "rss-folders-folderitem";
        if (filter.limit === "folder" && filter.id === folder._id) {
            className += " rss-folders-folderitem-active";
        }

        return (
            <div>
                <div
                    className={className}
                    onMouseEnter={this._handleShowOptions}
                    onMouseLeave={this._handleHideOptions}
                >
                    {toggleFeedsIcon}
                    {title}
                    {unread}
                    {options}
                </div>
                <div className="clear" />
                {listFeeds}
            </div>
        );
    }
}

export default FolderItemComponent;
