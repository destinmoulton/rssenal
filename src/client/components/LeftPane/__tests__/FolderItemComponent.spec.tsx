import { Map } from "immutable";
import * as React from "react";

import { shallow } from "enzyme";
import * as Types from "../../../types";
import FolderItemComponent from "../FolderItemComponent";
import feeds from "../../../../../test/data/immutable/feeds";
import folder from "../../../../../test/data/folder";

const FILTER = {
    id: folder._id,
    limit: "folder"
};

const UNREAD_MAP: Types.TUnreadMapFolders = Map([[folder._id, 20]]);
describe("<FolderItemComponent />", () => {
    it("renders and matches the snapshot", () => {
        const wrapper = shallow(
            <FolderItemComponent
                entriesRemoveRead={jest.fn()}
                folderDelete={jest.fn()}
                filterChange={jest.fn()}
                filterVisibleEntries={jest.fn()}
                editFeed={jest.fn()}
                editFolder={jest.fn()}
                folder={folder}
                feeds={feeds}
                filter={FILTER}
                unreadMapGroups={UNREAD_MAP}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
