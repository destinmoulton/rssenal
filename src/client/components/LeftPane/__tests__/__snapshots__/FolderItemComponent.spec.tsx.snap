// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FolderItemComponent /> renders and matches the snapshot 1`] = `
<div>
  <div
    className="rss-folders-folderitem rss-folders-folderitem-active"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="rss-folders-folderitem-icon"
    >
      <Icon
        as="i"
        name="caret right"
        onClick={[Function]}
      />
    </div>
    <div
      className="rss-folders-folderitem-title"
      onClick={[Function]}
    >
      Lifestyle
    </div>
    <div
      className="rss-folders-folderitem-unread"
      onClick={[Function]}
    >
       [20]
    </div>
  </div>
  <div
    className="clear"
  />
</div>
`;
